File: demos_code_J1_conception_objet/04_couplage_paiement/DemoCouplage.java
public class DemoCouplage {
    // Le contrat est exprimé dans le vocabulaire du client métier.
    interface PaiementPort {
        boolean payer(int cents);
    }

    // Simulation d'un SDK externe.
    static final class StripeClient {
        String charge(int cents) {
            System.out.println("[Stripe SDK simulé] charge " + cents + " cents");
            return "paid";
        }
    }

    // L'adaptateur traduit le SDK externe vers le contrat métier.
    static final class StripeAdapter implements PaiementPort {
        private final StripeClient stripe;

        StripeAdapter(StripeClient stripe) {
            this.stripe = stripe;
        }

        @Override
        public boolean payer(int cents) {
            return "paid".equals(stripe.charge(cents));
        }
    }

    // Fake pratique pour tester/démontrer sans fournisseur externe.
    static final class FakePaiement implements PaiementPort {
        private final boolean resultat;

        FakePaiement(boolean resultat) {
            this.resultat = resultat;
        }

        @Override
        public boolean payer(int cents) {
            System.out.println("[Fake] payer " + cents + " cents -> " + resultat);
            return resultat;
        }
    }

    static final class ValiderCommande {
        private final PaiementPort paiement;

        ValiderCommande(PaiementPort paiement) {
            this.paiement = paiement;
        }

        boolean executer(int totalCents) {
            return paiement.payer(totalCents);
        }
    }

    public static void main(String[] args) {
        ValiderCommande prod = new ValiderCommande(
            new StripeAdapter(new StripeClient())
        );
        System.out.println("Avec Stripe = " + prod.executer(4200));

        ValiderCommande test = new ValiderCommande(new FakePaiement(false));
        System.out.println("Avec Fake = " + test.executer(4200));
    }
}

File: Demo en Python/demos_code_J1_conception_objet_python/04_couplage_paiement/DemoCouplage.py
from abc import ABC, abstractmethod


# Le contrat est exprimé dans le vocabulaire du client métier.
class PaiementPort(ABC):
    @abstractmethod
    def payer(self, cents: int) -> bool:
        pass


# Simulation d'un SDK externe.
class StripeClient:
    def charge(self, cents: int) -> str:
        print(f"[Stripe SDK simulé] charge {cents} cents")
        return "paid"


# L'adaptateur traduit le SDK externe vers le contrat métier.
class StripeAdapter(PaiementPort):
    def __init__(self, stripe: StripeClient) -> None:
        self.stripe = stripe

    def payer(self, cents: int) -> bool:
        return self.stripe.charge(cents) == "paid"


# Fake pratique pour tester/démontrer sans fournisseur externe.
class FakePaiement(PaiementPort):
    def __init__(self, resultat: bool) -> None:
        self.resultat = resultat

    def payer(self, cents: int) -> bool:
        print(f"[Fake] payer {cents} cents -> {self.resultat}")
        return self.resultat


class ValiderCommande:
    def __init__(self, paiement: PaiementPort) -> None:
        self.paiement = paiement

    def executer(self, total_cents: int) -> bool:
        return self.paiement.payer(total_cents)


def main() -> None:
    prod = ValiderCommande(
        StripeAdapter(StripeClient())
    )
    print(f"Avec Stripe = {prod.executer(4200)}")

    test = ValiderCommande(FakePaiement(False))
    print(f"Avec Fake = {test.executer(4200)}")


if __name__ == "__main__":
    main()
